contracts/clients/src/OperationResults.ts

```diff
@@ -0,0 +1,110 @@
+import { BigNumber, ContractReceipt, utils } from "ethers";
+import { ActionData } from "./signer";
+
+type OperationResultError = {
+  actionIndex: BigNumber;
+  message: string;
+};
+
+export type OperationResult = {
+  walletAddress: string;
+  nonce: BigNumber;
+  actions: ActionData[];
+  success: Boolean;
+  results: string[];
+  error?: OperationResultError;
+};
+
+const getError = (
+  success: boolean,
+  results: string[],
+): OperationResultError | undefined => {
+  if (success) {
+    return undefined;
+  }
+
+  // Single event "WalletOperationProcessed(address indexed wallet, uint256 nonce, bool success, bytes[] results)"
+  // Get the first (only) result from "results" argument.
+  const [errorResult] = results;
+  // remove methodId (4bytes after 0x)
+  const errorArgBytesString = `0x${errorResult.substring(10)}`;
+  const errorString = utils.defaultAbiCoder.decode(
+    ["string"],
+    errorArgBytesString,
+  )[0]; // decoded bytes is a string of the action index that errored.
+
+  const splitErrorString = errorString.split(" - ");
+  if (splitErrorString.length !== 2) {
+    throw new Error("unexpected error message format");
+  }
+
+  return {
+    actionIndex: BigNumber.from(splitErrorString[0]),
+    message: splitErrorString[1],
+  };
+};
+
+/**
+ * Gets the results of operations (and actions) run through VerificationGateway.processBundle.
+ * Decodes unsuccessful operations into an error message and the index of the action that failed.
+ *
+ * @param transactionReceipt Transaction receipt from a VerificationGateway.processBundle transaction
+ * @returns An array of decoded operation results
+ */
+export const getOperationResults = (
+  transactionReceipt: ContractReceipt,
+): OperationResult[] => {
+  if (!transactionReceipt.events || !transactionReceipt.events.length) {
+    throw new Error(
+      `no events found in transaction ${transactionReceipt.transactionHash}`,
+    );
```

Comment:
What's the idea behind throwing an error instead of just returning an empty array in these cases? Isn't that a valid result?

Reply:
It was mainly to indicate to a user they may have passed in the wrong transaction. But I could likely do a better check on the method called and throw an error on that instead and return empty from here.

Checking the VG contract, it looks like you can technically submit a completely empty bundle, but I'm sure it would revert when it attempts to validate the empty sig.

Reply:
Thinking through this some more, we likely can't assume someone only will call `VG.processBundle` as they may go through a `BLSExpander` function instead. I will likely remove this check. I hope someone doesn't get too frustrated if they pass the wrong txn receipt in...